### DISCUSSION TITLE
arm build milvus-cpu-standalone，query very slow

### DISCUSSION THREAD

**Discussion by zzh928253398** [milvus-io/milvus#28284] [Q&A and General discussion]
I use milvus 2.2.0 to build arm version.

I created 5 containers using docker-compose, including these 5 services: etcd, mini, mysql, milvus server (query and insert services), and milvus engine.

I have created 22 indexes, Two of them have a relatively large number of collections, one is 1.6 million, and the other is 1.4 million. The number of remaining collections is relatively small, basically only around 10000. 
Query takes nearly 90 seconds。

40 records will be inserted per second, and the insertion speed is normal. But the query is slow. When the quantity is only 1.2 million, the query speed is still normal, returning in 2-3 seconds. However, when the quantity reaches 1.6 million, the query speed becomes super slow, taking 90 seconds.

I created the index using Python SDK. The parameter is
{"metric_type": L2, "index_type": "IVF-FLAT", "params": {"nlist": 2048}}. 
Would it be much faster if I had to modify it to HNSW？ But do I need to recreate the index?

Every time I query, I will load the collection once. May I ask if this is a very important reason for affecting time consumption?

I tried to modify several parameters in milvus.yaml according to the official website documentation, but the query did not become faster after the modifications. My modification method is to only change one item at a time:

1. Change indexCoord. minSegmentNumRowsToEnableIndex from 1024 to 512 (longer query time)

2. Change dataCoord.segment.maxSize from 512 to 2048 (longer query time)

3. Change queryNode. cache. memoryLimit from 2147483648 to 4294967296 (query time remains the same)

My machine has 48 cores and 128GB of memory. My operating system is openEuler and my machine is an arm server。

**Comment by zzh928253398** [comment]
and the Python code for creating the collection is as follows:
    def create_collection(self, collection_name):
        # Create milvus collection if not exists
        try:
            field1 = FieldSchema(name="id", dtype=DataType.INT64, description="int64", is_primary=True, auto_id=True)
            field2 = FieldSchema(name="embedding", dtype=DataType.FLOAT_VECTOR, description="float vector",
                                    dim=VECTOR_DIMENSION, is_primary=False)
            schema = CollectionSchema(fields=[field1, field2], description="collection description")
            self.collection = Collection(name=collection_name, schema=schema)
            LOGGER.debug(f"Create Milvus collection: {collection_name}")
            return "OK"
        except Exception as e:
            LOGGER.error(f"Failed create collection in Milvus: {e}")
            sys.exit(1)

and the Python code for creating the index is as follows:
    def create_index(self, collection_name):
        # Create IVF_FLAT index on milvus collection
        try:
            self.set_collection(collection_name)
            default_index = {"metric_type": METRIC_TYPE, "index_type": "IVF_FLAT", "params": {"nlist": 2048}}
            status = self.collection.create_index(field_name="embedding", index_params=default_index)
            if not status.code:
                LOGGER.debug(
                    f"Successfully create index in collection:{collection_name} with param:{default_index}")
                return status
            else:
                raise Exception(status.message)
        except Exception as e:
            LOGGER.error(f"Failed to create index: {e}")
            sys.exit(1)

**Comment by zzh928253398** [comment]
My query method is written as follows：
    def search_vectors(self, collection_name, vectors, top_k):
        # Search vector in milvus collection
        try:
            self.set_collection(collection_name)
            self.collection.load()
            search_params = {"metric_type": METRIC_TYPE, "params": {"nprobe": 16}}
            res = self.collection.search(vectors, anns_field="embedding", param=search_params, limit=top_k)
            LOGGER.debug(f"Successfully search in collection: {res}")
            return res
        except Exception as e:
            LOGGER.error(f"Failed to search vectors in Milvus: {e}")
            sys.exit(1)

**Reply by yhmo** [reply]
How many vectors in the "vectors"?

**Comment by zzh928253398** [comment]
@lhotari @shanghaikid @matrixji @sileht Can you provide me with some suggestions？

**Reply by xiaofan-luan** [reply]
try 2.2.14 maybe?
you should see a big performance difference

**Reply by zzh928253398** [reply]
I think the current data volume is only 1.6 million, and it is abnormal for the query time to be so long. I think we can first see which link went wrong, and if it's really not possible, we can consider upgrading.

**Comment by xiaofan-luan** [comment]
when you say"Query", does it mean query or search?

**Comment by xiaofan-luan** [comment]
collection number will be a big issue.
Milvus 2.2.x can not support 10000+ collections. which will consume too many cpus.

Any way you can use partition or parittion key feature instead?

**Reply by xiaofan-luan** [reply]
for 1.6milliion datas, we would expect data to be returned in 10-20 milliseconds.
few things you can do
1. please offer cluster logs and your query so we can take a look
2. try to upgrade to 2.2.15 because we fix many bugs and you may hit some bug
3. try to use pprof and perf tool to profiling. It would be great if you installed prometheus monitoring tools

**Reply by xiaofan-luan** [reply]
BTW, I think so far the code looks good to me.

1. How large is the topK?
2. What is the output field?
3. use export log script so we can get some detailed logs on why search is slow(Be sure you trigger one search)
